### DISCUSSION TITLE
Display what has Ctrl+Z/Ctrl+Y done in editor

### DISCUSSION THREAD

**Discussion by cihe13375** [ppy/osu#24157] [Ideas]
Currently there is no way to see what has done by undo/redo (Ctrl+Z/Y) unless the mapper is lucky (i.e. can directly see the object being modified). Consider the following scenario:
1) open a beatmap
2) change some objects
3) deselect all objects, seek to somewhere else in the timeline
4) Ctrl+Z

There is no indication on what Ctrl+Z did, or even if it was executed at all. This have several consequences:
1) the mapper may accidentally undo something without aware of it;
2) the mapper may press Ctrl+Z too few/many times then they want, and is not aware of that;
3) the mapper doesn't know that some operations cannot be undone (either due to feature limits or design), while they think the undone has succeed; or vice versa.

All of these can easily lead to undesired changes to the map, and it's super difficult to find them out (when this happens I usually quit the editor without saving and then re-enter it again, hoping that I didn't press Ctrl+S too frequently).
The situation can be improved if the editor can display what has Ctrl+Z done. It may be splited into several tasks, from simple to complicated:
1) show a hint on the screen when pressing Ctrl+Z or Ctrl+Y, and potentially how many continuous undo/redo has been performed (if (2) is not implemented);
2) add a simple message to the hint summarizing what has been reverted, e.g. "change the shape of slider at 01:47:593", or "rotate the object group 02:30:575 (1,2,3,4)".
3) provide undo/redo history lists like Photoshop or Word, and the mapper may revert the map to any specific point in the lists.

**Comment by peppy** [comment]
This shouldn't be too hard to add. Interestingly, stable kinda did this internally (but was never displayed).
